lib/stm32/l1/flash.c

```diff
@@ -230,5 +230,50 @@ void eeprom_program_words(uint32_t address, uint32_t *data, int length_in_words)
 	flash_lock_pecr();
 }
 
+/** @brief Erase a page (256 Byte) of program memory (flash)
+ *
+ * @param[in] page_address must point to flash space, no checking!
+ */
+void flash_erase_page(uint32_t address)
+{
+	while (FLASH_SR & FLASH_SR_BSY);
+	FLASH_PECR |= (FLASH_PECR_ERASE | FLASH_PECR_PROG);
+	*(uint32_t*)address = 0;
+	while (FLASH_SR & FLASH_SR_BSY);
+	FLASH_PECR &= ~(FLASH_PECR_ERASE | FLASH_PECR_PROG);
+}
+
+/** @brief Program memory: write half page to flash
+ *
+ * A half page is 32 words (128 byte).
+ *
+ * Half Page Write is possible only from SRAM, from external memory or from
+ * bank 1 to bank 2 and vice versa. Force into SRAM with the section attribute.
+ *
+ * @param[in] address must point to flash space, no checking!
+ * @param[in] data pointer to 128 bytes of data to write
+ */
+__attribute__((section(".data#")))
+void flash_program_half_page(uint32_t address, uint32_t *data)
+{
+	uint32_t* p = (uint32_t*)address;
+	FLASH_PECR |= (FLASH_PECR_FPRG | FLASH_PECR_PROG);
+	while (FLASH_SR & FLASH_SR_BSY);
+	for (int i = 0; i < 32; i++)
+		*p++ = *data++;
+	while (FLASH_SR & FLASH_SR_BSY);
+	FLASH_PECR &= ~(FLASH_PECR_FPRG | FLASH_PECR_PROG);
+}
+
+/** @brief Program memory: write one word
+ *
+ * @param[in] address must point to flash space, no checking!
+ * @param[in] data word to write
+ */
+void flash_program_word(uint32_t address, uint32_t data)
+{
+	while (FLASH_SR & FLASH_SR_BSY);
+	*(uint32_t*)address = data;
```

Comment:
Also this is similar to other flash operations e.g. in flash_common_f24.c `flash_program_half_word()` with the difference that there it is in a function `flash_wait_for_last_operation()`.

About locking/unlocking I reckon this should be done separately, since usually you'd want to write more than one word. Also it is similar to `flash_program_half_word()` in flash_common_f24.c